sdk/include/opentelemetry/sdk/instrumentationscope/scope_configurator.h

```diff
@@ -0,0 +1,141 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+#include <functional>
+
+#include "opentelemetry/sdk/instrumentationscope/instrumentation_scope.h"
+#include "opentelemetry/version.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace instrumentationscope
+{
+/**
+ * A scope configurator is a function that returns the scope config for a given instrumentation
+ * scope.
+ */
+template <typename T>
+class ScopeConfigurator
+{
+public:
+  /**
+   * A builder class for the ScopeConfigurator that facilitates the creation of ScopeConfigurators.
+   */
+  class Builder
+  {
+  public:
+    /**
+     * Constructor for a builder object that cam be used to create a scope configurator. A minimally
+     * configured builder would build a ScopeConfigurator that applies the default_scope_config to
+     * every instrumentation scope.
+     * @param default_scope_config The default scope config that the built configurator should fall
+     * back on.
+     */
+    explicit Builder(T default_scope_config) noexcept : default_scope_config_(default_scope_config)
+    {}
+
+    /**
+     * Allows the user to pass a generic function that evaluates an instrumentation scope through a
+     * boolean check. If the check passes, the provided config is applied. Conditions are evaluated
+     * in order.
+     * @param scope_matcher a function that returns true if the scope being evaluated matches the
+     * criteria defined by the function.
+     * @param scope_config the scope configuration to return for the matched scope.
+     * @return this
+     */
+    Builder AddCondition(std::function<bool(const InstrumentationScope &)> scope_matcher,
+                         T scope_config)
+    {
+      conditions_.push_back(Condition{scope_matcher, scope_config});
+      return *this;
+    }
+
+    /**
+     * A convenience condition that specifically matches the scope name of the scope being
+     * evaluated. If the scope name matches to the provided string, then the provided scope
+     * configuration is applied to the scope.
+     * @param scope_name The scope name to which the config needs to be applied.
+     * @param scope_config The scope config for the matching scopes.
+     * @return this
+     */
+    Builder AddConditionNameEquals(nostd::string_view scope_name, T scope_config)
+    {
+      std::function<bool(const InstrumentationScope &)> name_equals_matcher =
+          [scope_name](const InstrumentationScope &scope_info) {
```

Comment:
Passing `string_view` in the context looks dangerous, because one is not guaranteed that the contents of scope_name will still be around when the matcher is invoked. One possibility would be to own a copy

```suggestion
          [scope_name=std::string(scope_name)](const InstrumentationScope &scope_info) {
```